### PR TITLE
Data filter

### LINKED ISSUE #233: OpenSDS Controller API Request Filter
**Is this a BUG REPORT or FEATURE REQUEST?**:
> /kind feature

**What happened**:
Standard methods for REST are List, Get, Create, Update and Delete. Specially,The List method takes a collection name and zero or more parameters as input and returns a list of resources that match the input. In the current version , the OpenSDS Restfull API doesn't support the paination, order sorting, criteria query, total count of items obatainment.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Hotpot(release/branch) version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by lasshpanzer**
**What this PR does / why we need it**:
This PR adds supports to query resources filtering by properties.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #233 
The design spec is drafted [here](https://docs.google.com/document/d/1f-ta2CbORuXPTdEZybl48hqbBpq2wnsy1puxbzlvLik/edit#).
**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/db/db.go [hunk 1]
```diff
@@ -58,7 +57,9 @@ type Client interface {
 
 	GetDock(dckID string) (*model.DockSpec, error)
 
-	ListDocks() ([]*model.DockSpec, error)
+	GetDocks() ([]*model.DockSpec, error)
```
Comment (leonwanghui) [inline pkg/db/db.go]: Can you keep ListDocks method the same with before, and add a new method(for example `ListDocksWithFilter` or `ListFilteredDocks` such stuffs), in this way it should reduce some code changes.

**Comment by xing-yang** [comment]
In the design spec https://docs.google.com/document/d/1f-ta2CbORuXPTdEZybl48hqbBpq2wnsy1puxbzlvLik/edit#, there is a filter volumes by “lun_id”. This is for admin-only.  I don’t see that in this PR.  "lun_id" is the id of the lun on the storage backend.  If this can't be addressed by this PR yet, please add a TODO note for it to be addressed later.

Also it looks like this patch will accept any filter.  There is no validation on what is supported and what is not supported.  Please add validation based on the filters listed in the design doc.  Thanks.

File: pkg/db/drivers/etcd/etcd.go [hunk 2]
```diff
@@ -561,6 +991,127 @@ func (c *Client) ListVolumes() ([]*model.VolumeSpec, error) {
...
+
+	case "SIZE":
+		return volume[i].Size < volume[j].Size
+
+	case "POOLID":
+		return volume[i].PoolId < volume[j].PoolId
+
+	case "DESCRIPTION":
+		return volume[i].Description < volume[j].Description
+	}
```
Comment (xing-yang) [inline pkg/db/drivers/etcd/etcd.go]: Please add a TODO note for LunId and GroupId.

File: pkg/db/drivers/etcd/etcd.go [hunk 3]
```diff
@@ -561,6 +991,127 @@ func (c *Client) ListVolumes() ([]*model.VolumeSpec, error) {
...
+	case "STATUS":
+		return volume[i].Status < volume[j].Status
+
+	case "AVAILABILITYZONE":
+		return volume[i].AvailabilityZone < volume[j].AvailabilityZone
+
+	case "PROFILEID":
+		return volume[i].ProfileId < volume[j].ProfileId
+
+	case "PROJECTID":
```
Comment (xing-yang) [inline pkg/db/drivers/etcd/etcd.go]: Please use TENANTID instead.

We need to replace all ProjectId with TenantId as a bug fix.  Not in this patch.  But please use TenantId when you create new parameters.

File: pkg/db/drivers/etcd/etcd.go [hunk 4]
```diff
@@ -862,6 +1517,113 @@ func (c *Client) ListVolumeSnapshots() ([]*model.VolumeSnapshotSpec, error) {
...
+
+	case "STATUS":
+		return volumeSnapshot[i].Status < volumeSnapshot[j].Status
+
+	case "USERID":
+		return volumeSnapshot[i].UserId < volumeSnapshot[j].UserId
+
+	case "PROJECTID":
+		return volumeSnapshot[i].ProjectId < volumeSnapshot[j].ProjectId
+	}
```
Comment (xing-yang) [inline pkg/db/drivers/etcd/etcd.go]: VolumeSize?

Comment (xing-yang) [reply pkg/db/drivers/etcd/etcd.go]: Add a TODO note for GroupSnapshotId.

Comment (lasshpanzer) [reply pkg/db/drivers/etcd/etcd.go]: All ProjectId  will be modified by another PR

**Comment by leonwanghui** [comment]
@lasshpanzer There are some conflicts with current code, please rebase it to update.

**Comment by leonwanghui** [comment]
What's the relationship between this patch and #298  ?

**Pull request closed without merging**